pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java

```diff
@@ -151,4 +148,18 @@ private void flush()
       LOG.info("Pinot segment uploaded to {}", segmentURI);
     });
   }
+
+  @Override
+  public List<GenericRow> snapshotState(long checkpointId, long timestamp) {
```

Comment:
How often this would be called from the Flink side? Do we incrementally call this periodically over small `_pendingRows`?

Unless this snapshotState() gets called very frequently (i.e. getting called for each input row), I think that storing all rows in memory can consume the memory pretty quickly and we can easily go OOM when there are many columns.

Reply:
It depends on the configured checkpointing interval. It is called during every checkpoint. I was concerned about storing the rows in memory as well. I think a better solution would be to store them in flink state as they come in, then they can be stored in rocksdb. I can make that change if you concur @snleee ?

Reply:
@bobby-richard  After giving a bit of thoughts, I would like to understand what you are trying to handle here. The simplest way of handling fault tolerant is to depend on something like the following:

1. when flush() gets called, we update some state (e.g. checkpoint = x) to indicate "all data up to `checkpoint` has been flushed to Pinot. (In Kafka, the checkpoint can be the offset).
2. if the flink job fails in the middle, we start to consume from `x+1`.

I think that it's inefficient to store all the rows as the checkpoint because we simply need to store the `last offset that we flushed in the last time`. Do you think that the above mechanism is feasible on Flink?

@walterddr Can you also chime in since you have the experience with the Flink?

Reply:
@snleee Let's say we are using a kafka source to feed the pinot sink with checkpointing enabled at a 1 minute interval. Every minute when the checkpoint is taken, the kafka source will persist its current offsets in the checkpoint. If the job fails, it will be restarted from the last checkpoint. The kafka source will resume from the last committed offsets, but the pinot sink will have lost all pending rows for the current unflushed segment up until that point. 

You can rewind data from the source if running in batch mode without checkpointing enabled. But if you are running a streaming job with a kafka source, it's impossible to match up the kafka offsets of the source with the last row of the most recently flushed segment.

I'm not sure how you can support checkpointing in the sink without storing the pending rows in state.

Reply:
I see. Have you thought of using Kafka Sink on Flink side and let Pinot directly consume from the output kafka topic? Pinot's real-time consumption handles the offset commit by itself and it handles fault tolerance by resuming to consume from the offset that has been committed recently on the Pinot side.

For this PR, let's at least go with your 2nd approach using Flink state to unblock you. In case of Flink state, does Flink take care of cleaning up the RocksDB data? If not, we should handle the data clean up when exiting the `flush()`.

Reply:
And yes, the state would need to be cleared during the flush method. Could also only store the pending rows in state if checkpointing is enabled so there is zero penalty if checkpointing is not enabled.

Reply:
@bobby-richard Very interesting use case 👍 In that case, in-memory solution may satisfy your requirements given that the rows that would flow to the flink job would be very small.

RocksDB solution may requires extra complexity because we need to take care of state clean-up and we should also take care of failure cases (e.g. the node gets killed while we are cleaning up the rocks db etc).

We can also take the incremental approach (supporting in-memory mode for now and we can add the rocksdb mode for checkpoint storage).

Reply:
Flink should manage all of the RocksDB complexity for us. Just clearing the state at the appropriate times should be enough. I really think the state solution is the way to go, otherwise the job will require very large heap memory allocations. 

Time based trigger for flush is interesting. I assume that would be configurable? Storing the data in flink state for a very long time is really no problem though.

Reply:
PR updated to use flink state